### PR TITLE
HIVE-26436: Hive on MR NullPointerException When initializeOp has not…

### PR DESCRIPTION
**Pull request by ganlute**
… been called and close called.

### What changes were proposed in this pull request?
When the operator has not been called to initialize, and it is called to close, exception will happen.
I fix it by check conf is null so that the close could be finished. It work well in my cluster. 
However, I am not sure, it may be a better way to skip close if the operator has not been initialized ?

### Why are the changes needed?
I think it may be a bug.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
1. unit test passed.
2. apply the patch to cluster and task passed.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 1]
```diff
@@ -3770,7 +3770,7 @@ private static Path createDummyFileForEmptyTable(JobConf job, MapWork work,
   /**
    * setInputPaths add all the paths in the provided list to the Job conf object
    * as input paths for the job.
-   *
+   *
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:3773]: Unrelated change.

Comment (ganlute) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:3773]: > 

I will remove it since I add it for triggering CI

**Comment by ganlute** [comment]
> Can you add a unit test as well for this change

OK, I will update the unit test.

**Comment by ganlute** [comment]
hello @ayushtkn, thanks for your suggestion, I have add unit test to it. 
BTW, I found that not all the operator has the problem and it can happen at FileSinkOperator so I add it to the test case.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Operator.java [hunk 2]
```diff
@@ -675,6 +678,13 @@ public void close(boolean abort) throws HiveException {
       return;
     }
 
+    // if the operator has not been initialized, there is no need to close
+    if (!this.rootInitializeCalled) {
+      LOG.debug("No need to close operator {}", this);
+      return;
+    }
+
+
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Operator.java:687]: why this is specific to MR and doesn't happen with Tez? And what is the actual scenario/query where this happens, the extended test doesn't seems to be very indicative

Comment (ganlute) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Operator.java:687]: > why this is specific to MR and doesn't happen with Tez? And what is the actual scenario/query where this happens, the extended test doesn't seems to be very indicative

@anishek I am so sorry for the slow reply. 

In fact, I can not reproduce the problem. It happened in one of my online clusters and I can not reproduce it in my new cluster. Some of configs may be different but I am not sure which one cause the problem.

Here is the map file printed when exception happened:
<MAP>Id =0
  <Children>
    <TS>Id =156
      <Children>
        <UNION>Id =26
          <Children>
            <SEL>Id =40
              <Children>
                <FS>Id =146
                  <Children>
                  <\Children>
                  <Parent>Id = 40 null<\Parent>
                <\FS>
              <\Children>
              <Parent>Id = 26 null<\Parent>
            <\SEL>
          <\Children>
          <Parent>Id = 156 nullId = 162 
        <TS>Id =162
          <Children>null
          <\Children>
          <Parent><\Parent>
        <\TS>Id = 166 
        <TS>Id =166
          <Children>null
          <\Children>
          <Parent><\Parent>
        <\TS><\Parent>
        <\UNION>
      <\Children>
      <Parent>Id = 0 null<\Parent>
    <\TS>
  <\Children>
  <Parent><\Parent>
<\MAP>

However, I think for whatever reason the operator is not initialized, close should not throw an exception.
